### PR TITLE
feature/cs-216-multilingual-agent

### PR DESCRIPTION
**Pull request by ThWoywod**
Add the ability to generate Dialogflow compliant multilingual agents.

### CODE REVIEW

File: src/components/apiai/generator.ts [hunk 1]
```diff
@@ -3,55 +3,67 @@ import { GenericIntent, PlatformGenerator } from "assistant-source";
...
     console.log("=============     PROCESSING ON APIAI     ============");
-    console.log("Intents: #" + intentConfigurations.length + ", language: " + language);
+    console.log(`Intents: #${intentConfigurations.length}, languages: ${languages.join(",")}`);
 
     console.log("validating...");
-    const convertedIntents = this.prepareConfiguration(intentConfigurations);
 
-    console.log("building entities (" + Object.keys(customEntityMapping).length + ")...");
+    const defaultLanguage = this.component.configuration.defaultLanguage;
+    if (!intentConfigurations[defaultLanguage]) throw new Error(`Missing intent configuration for language "${defaultLanguage}"`);
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: please change to:

```ts
throw new Error(`Missing intent configuration for configured default language "${defaultLanguage}". You may change the default language in your apiai component configuration.`);
```

File: src/components/apiai/generator.ts [hunk 2]
```diff
@@ -3,55 +3,67 @@ import { GenericIntent, PlatformGenerator } from "assistant-source";
...
+
+    const convertedIntents = this.languages
+      .map(language => {
+        return { [language]: this.prepareConfiguration(intentConfigurations[language]) };
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: You can omit the `return` with this: 

``.map(language => ({[language]: ...})`

File: src/components/apiai/generator.ts [hunk 3]
```diff
@@ -3,55 +3,67 @@ import { GenericIntent, PlatformGenerator } from "assistant-source";
...
+      })
+      .reduce((previousValue, currentValue) => {
+        return { ...previousValue, ...currentValue };
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: You should omit the return:

``
.reduce((prev, curr) => ({...prev, ...cur}))
``

File: src/components/apiai/generator.ts [hunk 4]
```diff
@@ -63,37 +75,49 @@ export class Generator implements PlatformGenerator.Extension {
...
+            }
+          }),
         };
-        const entries = customEntityMapping[type];
-        return { entity, entries };
-      }
-    });
+      })
+      .reduce((previousValue, currentValue) => {
+        return { ...previousValue, ...currentValue };
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: see above regarding the `return`

File: src/components/apiai/generator.ts [hunk 5]
```diff
@@ -63,37 +75,49 @@ export class Generator implements PlatformGenerator.Extension {
...
-          id: uuid(),
-          name: type,
-          isOverridable: true,
-          isEnum: false,
-          automatedExpansion: false,
+    return this.languages
+      .map(language => {
+        return {
+          [language]: Object.keys(customEntityMapping[language]).map(type => {
+            if (typeof config.entities[type] !== "undefined") {
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: what happens if `config.entities[type]` is `undefined`? Do we need a warning in this case?

File: src/components/apiai/generator.ts [hunk 6]
```diff
@@ -117,7 +141,20 @@ export class Generator implements PlatformGenerator.Extension {
         events: this.prepareEventsFor(config.intent),
       };
 
-      const utterances = config.utterances.map(utterance => this.buildUtterance(utterance, entityMapping, customEntityMapping));
+      const utterances = this.languages
+        .map(language => {
+          const currentLanguageConfiguration = preparedIntentConfiguration[language].find(currentConfig => currentConfig.intent === config.intent);
+          if (currentLanguageConfiguration) {
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: what happens if we can't find a `currentLangaugeConfiguration`? Do we need a warning in this case? Cause your `map` will result to `undefined` then.

File: src/components/apiai/generator.ts [hunk 7]
```diff
@@ -277,39 +331,58 @@ export class Generator implements PlatformGenerator.Extension {
...
+  /**
+   * Writes generated custom entities to entity definition and language specific synonym file
+   * @param customEntity
+   */
+  private writeCustomEntitiesFiles(customEntities) {
+    const defaultCustomEntities = customEntities[this.component.configuration.defaultLanguage];
+    defaultCustomEntities.forEach(defaultCustomEntity => {
+      if (typeof defaultCustomEntity !== "undefined") {
+        fs.writeFileSync(`${this.entitiesDirectory}/${defaultCustomEntity.entity.name}.json`, JSON.stringify(defaultCustomEntity.entity, null, 2));
+        this.languages.forEach(language => {
```
Comment (antoniusostermann) [inline src/components/apiai/generator.ts]: please add a whitespace here

File: src/components/apiai/generator.ts (hunk 6 above)
Comment (ThWoywod) [reply src/components/apiai/generator.ts]: The `reduce`-function could handle this and all `undefined` values will be removed. So no error will appear. But it's a good hint to add a warning massage.  👍

**Pull request merged**